### ISSUE TITLE
Submessages are read only by default

### ISSUE THREAD

**Issue by pauldemarco** [google/protobuf.dart#305]
Sample proto:
```proto
message Person {
    Address address = 1;
}

message Address {
    string city = 1;
}
```

Test:
```dart
Person person = Person();
person.address.city = "Hello";
print(person);
```

Expected:
```log
address: {
  city: Hello
}
```

Actual:
```log
Unsupported operation: Attempted to change a read-only message (Address)
package:protobuf/src/protobuf/field_set.dart 15:3            defaultFrozenMessageModificationHandler
package:protobuf/src/protobuf/field_set.dart 157:54          _FieldSet._ensureWritable
package:protobuf/src/protobuf/field_set.dart 794:5           _FieldSet._validateField
package:protobuf/src/protobuf/field_set.dart 512:5           _FieldSet._$check
package:protobuf/src/protobuf/field_set.dart 493:12          _FieldSet._$set
package:protobuf/src/protobuf/generated_message.dart 457:58  GeneratedMessage.$_setString
package:todo_list/gen/app_state.pb.dart 306:30               Address.city=
test/widget_test.dart 18:20                                  main.<fn>
```

I was hoping all fields of the protobuf message (at any level) would be writable up until a `freeze()` is called, is this intended operation?

**Comment by jasonkolodziej** [comment]
@pauldemarco Were you able to find a solution to this?

**Comment by w4lle** [comment]
Has any solutions?

**Comment by sigurdm** [comment]
The issue is that `person.address` returns an immutable 'default value'.
This is a pretty clumsy design but a bit hard to change now.

The workaround is: `person.address = (Address()..city = 'Boston')`

**Comment by w4lle** [comment]
@sigmundch It can work, thanks.

**Comment by pauldemarco** [comment]
@sigurdm thanks for clarifying. Let’s consider **Address** had a bunch of properties but I only wanted to chance the city. What would be the least verbose way of changing just one inner field?

**Comment by sigurdm** [comment]
We recently added the ensureX methods
```
person.ensureAdress.city = 'boston';
```

Does that work?

**Comment by pauldemarco** [comment]
@sigurdm Like you've mentioned, this is only an issue when Person is created without an explicit Address as well:
```dart
final person = Person();
person.address.city = "lancaster";  // read-only error
```
vs.
```dart
final person = Person()..address=Address();
person.address.city = "lancaster" // no error
```

I haven't done any deep dives into the generated code yet. But is there anyway to create a Person without all of these immutable default values within?

**Comment by sigurdm** [comment]
>I haven't done any deep dives into the generated code yet. Is there anyway to create a Person that has mutable default values within?

No, I don't think so.

Can you explain the use-case? Maybe there are alternatives...

**Comment by osa1** [comment]
The code in the original example is incorrect and cannot be fixed in the library. The real problem here is #309. Let's consider the original example:

```proto
syntax = 'proto3';

message Person {
    Address address = 1;
}

message Address {
    string city = 1;
}
```

When we create a `Person` and don't set the `address` field, `hasAddress()` returns false as expected:

```dart
Person p = Person.create();
print(p.hasAddress()); // false
```

If we want `p.address.city = ...` to work, in the getter `p.address` we need to set this field. That means just getting the field will change the presence of this field. So if we implement the requested behavior is this issue this is what we would observe:

```dart
Person p = Person.create();
print(p.hasAddress()); // false
print(p.address);
print(p.hasAddress()); // true!!!
```

The getter needs to set the field to make `p.address.city = ...` work, but that means getter changes presence of the field.

Instead we should fix #309 and return `null` when a message field is not set.

Closing this in favor or #309.
